File: CompositePattern/python/Menu.py
import abc
from typing import *

class MenuItem(metaclass=abc.ABCMeta):
    def __init__(self, description:str) -> None:
        self._description:str = description

    @abc.abstractmethod
    def __str__(self) -> str:
        return NotImplemented
    

class Menu(MenuItem):
    def __init__(self, description:str) -> None:
        super().__init__(description)
        self._components:List[MenuItem] = list()

    def addMenuItem(self, menuItem:MenuItem):
        self._components.append(menuItem)
    
    def removeMenuItem(self, menuItem:MenuItem) -> bool:
        try:
            self._components.remove(menuItem)
            return True
        except ValueError:
            return False
    

class MenuBook(Menu):
    def __init__(self, description:str) -> None:
        super().__init__(description)

    def __str__(self) -> str:
        output = f"======    {self._description}    ======\n"
        for component in self._components:
            output += f"{str(component)}\n"

        return output
    

class SubMenu(Menu):
    def __init__(self, description:str) -> None:
        super().__init__(description)

    def __str__(self) -> str:
        output = f"SubMenu {self._description}:\n"
        for component in self._components:
            if isinstance(component, Item):
                output += f"\t{str(component)}\n"
            else:
                temp = str(component).replace('\n', '\n\t')
                output += f"\n\t{temp}"

        return output
    

class Item(MenuItem):
    def __init__(self, description:str) -> None:
        super().__init__(description)

    def __str__(self) -> str:
        return self._description
    

File: CompositePattern/cpp/src/menu.cpp
#include <queue>
#include "header/menu.h"


MenuItem::MenuItem(string description){
    this->description = description;
}

Menu::Menu(string description): MenuItem(description){};

void Menu::addMenuItem(shared_ptr<MenuItem> &menuItem){
    components.push_back(menuItem);
}

void Menu::addMenuItem(shared_ptr<MenuItem> &&menuItem){
    components.push_back(menuItem);
}

bool Menu::removeMenuItem(shared_ptr<MenuItem> &menuItem){
    for(auto it=components.begin(); it != components.end(); it++){
        if(*it == menuItem){
            components.erase(it);
            return true;
        }
    }

    return false;
}

bool Menu::removeMenuItem(shared_ptr<MenuItem> &&menuItem){
    for(auto it=components.begin(); it != components.end(); it++){
        if(*it == menuItem){
            components.erase(it);
            return true;
        }
    }

    return false;
}

bool Menu::isItem(){
    return false;
}

MenuBook::MenuBook(string description): Menu(description){};

string MenuBook::to_string(){
    string output = "======    " + description + "    ======\n";
    for(auto component: components){
        output += (component->to_string() + "\n");
    }

    return output;
}

SubMenu::SubMenu(string description): Menu(description){};

string SubMenu::to_string(){
    string output = "SubMenu " + description +":\n";
    for(auto component: components){
        if(component->isItem()){
            output += ("\t" + component->to_string() + "\n");
        }else{
            string s = component->to_string();
            int pos = s.find("\n", 0);
            while(pos != string::npos){
                s.replace(pos, 1, "\n\t");
                pos = s.find("\n", pos+2);
            }
            output += ("\n\t" + s);
        }
    }

    return output;
}

Item::Item(string description): MenuItem(description){};


bool Item::isItem(){
    return true;
}

string Item::to_string(){
    return description;
}
